File: apps/api/app/ml/splitting/split.py
"""Deterministic, reproducible train/test splitting."""

from typing import Any, TypeVar
import numpy as np

T = TypeVar("T")


def train_test_split(
    data: list[T],
    train_ratio: float = 0.8,
    seed: int = 42,
) -> tuple[list[T], list[T]]:
    """Split data into disjoint train and test subsets deterministically.

    Guarantees:
    - Partition: train and test indices are disjoint and span the whole dataset.
    - Determinism: identical outputs given the same seed and ratio.
    """
    n = len(data)
    if n < 2:
        raise ValueError("Data must contain at least 2 samples to split into train and test")
    if not (0.0 < train_ratio < 1.0):
        raise ValueError(f"train_ratio must be strictly between 0.0 and 1.0, got {train_ratio}")

    rng = np.random.default_rng(seed)
    shuffled_indices = rng.permutation(n)

    train_count = int(round(train_ratio * n))
    # Ensure at least 1 sample in each partition
    train_count = max(1, min(n - 1, train_count))

    train_indices = set(shuffled_indices[:train_count])
    test_indices = set(shuffled_indices[train_count:])

    # Verify no overlap
    assert len(train_indices.intersection(test_indices)) == 0, "Train and test partitions overlap"

    train_data = [data[i] for i in shuffled_indices[:train_count]]
    test_data = [data[i] for i in shuffled_indices[train_count:]]

    return train_data, test_data

File: apps/web/features/workbench/split.ts
import type { SplitConfig } from "./types";

/**
 * 32-bit seeded pseudo-random number generator (Mulberry32).
 */
function createSeededRandom(seed: number): () => number {
  let a = seed >>> 0;
  return function () {
    let t = (a += 0x6d2b79f5);
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

export function splitTrainTest<T>(
  items: T[],
  config: SplitConfig
): { train: T[]; test: T[] } {
  const n = items.length;
  if (n < 2) {
    throw new Error("Dataset must contain at least 2 samples to split");
  }
  if (config.trainRatio <= 0 || config.trainRatio >= 1) {
    throw new Error("trainRatio must be strictly between 0 and 1");
  }

  const rng = createSeededRandom(config.seed);

  // Fisher-Yates shuffle on indices
  const indices = Array.from({ length: n }, (_, i) => i);
  for (let i = n - 1; i > 0; i--) {
    const j = Math.floor(rng() * (i + 1));
    const temp = indices[i];
    indices[i] = indices[j];
    indices[j] = temp;
  }

  const trainCount = Math.max(1, Math.min(n - 1, Math.round(config.trainRatio * n)));

  const train: T[] = [];
  const test: T[] = [];

  indices.forEach((idx, pos) => {
    if (pos < trainCount) {
      train.push(items[idx]);
    } else {
      test.push(items[idx]);
    }
  });

  return { train, test };
}
